advanced_source/ddp_pipeline_tutorial.py

```diff
@@ -0,0 +1,452 @@
+    val_data = batchify(val_txt, eval_batch_size, rank, world_size)
+    test_data = batchify(test_txt, eval_batch_size, rank, world_size)
+
+
+######################################################################
+# Functions to generate input and target sequence
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+
+
+######################################################################
+# ``get_batch()`` function generates the input and target sequence for
+# the transformer model. It subdivides the source data into chunks of
+# length ``bptt``. For the language modeling task, the model needs the
+# following words as ``Target``. For example, with a ``bptt`` value of 2,
+# we’d get the following two Variables for ``i`` = 0:
+#
+# .. image:: ../_static/img/transformer_input_target.png
+#
+# It should be noted that the chunks are along dimension 0, consistent
+# with the ``S`` dimension in the Transformer model. The batch dimension
+# ``N`` is along dimension 1.
+#
+
+# In 'run_worker'
+    bptt = 35
+    def get_batch(source, i):
+        seq_len = min(bptt, len(source) - 1 - i)
+        data = source[i:i+seq_len]
+        target = source[i+1:i+1+seq_len].view(-1)
+        return data, target
+
+######################################################################
+# Model scale and Pipe initialization
+# -----------------------------------
+#
+
+
+######################################################################
+# To demonstrate training large Transformer models using pipeline parallelism,
+# we scale up the Transformer layers appropriately. We use an embedding
+# dimension of 4096, hidden size of 4096, 16 attention heads and 8 total
+# transformer layers (``nn.TransformerEncoderLayer``). This creates a model with
+# **~1 billion** parameters.
+#
+# We need to initialize the `RPC Framework <https://pytorch.org/docs/stable/rpc.html>`__
+# since Pipe depends on the RPC framework via `RRef <https://pytorch.org/docs/stable/rpc.html#rref>`__
+# which allows for future expansion to cross host pipelining. We need to
+# initialize the RPC framework with only a single worker since we're using a
+# single process to drive multiple GPUs.
+#
+# The pipeline is then initialized with 8 transformer layers on one GPU and 8
+# transformer layers on the other GPU. One pipe is setup across GPUs 0 and 1 and
+# another across GPUs 2 and 3. Both pipes are then replicated using DistributedDataParallel.
+
+# In 'run_worker'
+    ntokens = len(TEXT.vocab.stoi) # the size of vocabulary
+    emsize = 4096 # embedding dimension
+    nhid = 4096 # the dimension of the feedforward network model in nn.TransformerEncoder
+    nlayers = 8 # the number of nn.TransformerEncoderLayer in nn.TransformerEncoder
+    nhead = 16 # the number of heads in the multiheadattention models
+    dropout = 0.2 # the dropout value
+
+    from torch.distributed import rpc
+    tmpfile = tempfile.NamedTemporaryFile()
```

Comment:
So do both processes initialize an independent instance of the RPC framework (i.e. both processes launched by run_worker have RPC initialized, but they're not interconnected?) Is it possible for only one process to init RPC?

Reply:
Both processes need to initialize RPC since both are driving independent pipes. The RPC framework across both don't have to be interconnected since these are two independent pipes. When we do support cross host pipelining, we need RPCs only across a single instance of the pipe.

In this particular example, I didn't create an RPC worker group of world_size 2 since it might be misleading that we're performing RPCs across two processes.